044-cluster-wide-volume-usage-quota-management.md

```diff
@@ -0,0 +1,328 @@
+replication listener of the broker.
+
+##### Limit Type Configuration
+
+The existing limit types applied to the aggregate used bytes of all volumes will be meaningless now that we are
+working with a description of all volumes in the cluster. Nodes enter and exit the active set as part of normal operation
+and so the aggregate used bytes will vary.
+
+Instead, we propose introducing new limit types applied on a per-volume basis. Meaning that there is a single value
+for each limit which we test against each volume. i.e. we do not support limiting based on a specific volume. When
+that limit is exceeded we effectively stop messages being produced (by calculating a Throttle Factor of 0.0).
+
+We will not support the pre-existing concept of a **soft limit**. In the existing code the soft limit is where
+you begin slowing down the rate of message production. As you approach the hard limit the production quota is reduced
+linearly. Then at the hard limit message production is stopped. We decided this additional signal that something is
+failing is not a core feature of this plugin. We expect users to have their own disk usage monitoring to prompt
+intervention before this plugin shuts down all message production.
+
+We require one limit to be configured.
+
+Defining multiple limit types would be an invalid state.
+
+The limits we want are:
+1. throttle if [availableBytes](#volume) is less-than-or-equal-to some threshold for any volume
+2. throttle if the [availableRatio](#volume) is less-than-or-equal-to some threshold for any volume
+
+For example, to configure a limit when availableBytes is below 1GB:
+- `client.quota.callback.static.storage.per.volume.limit.min.available.bytes=1000000000`
+
+Another example, to configure a limit when availableRatio is below 0.01 (1%):
+- `client.quota.callback.static.storage.per.volume.limit.min.available.ratio=0.01`
+
+Defining more than one limit is **invalid**, for example the below is disallowed:
+- `client.quota.callback.static.storage.per.volume.limit.min.available.ratio=0.05`
+- `client.quota.callback.static.storage.per.volume.limit.min.available.bytes=5000000000`
+
+#### Throttle Factor Fallback
+
+We are going to depend on using the admin client to get volume information. This brings all the baggage of making a
+connection and dealing with possible failures. Also, we can potentially have an inconsistent view of the world between
+determining the active broker set and asking for a description of all log dirs (because we make two independent API calls)
+. So we need to react somehow to cases where we cannot get the volume data for all active brokers.
+
+Example inconsistent state:
+1. we call `describeCluster` and get a response that says broker 1 and 2 are active
+2. broker 2 shuts down cleanly and is removed from the active set
+3. we call `describeLogDirs( brokerIds = [1,2] )` and only receive descriptions for logdirs on broker 1
```

Comment:
in the example above, is the plugin going to use fallback immediately or do you envisage at least a "number" of retries to get consistent information based on ephemeral communication errors between admin client and cluster?
PS: I know there is a throttle factor validity duration used for transient failures anyway.

Reply:
Yes, I imagine we'll also implement a naive retry when we fetch the cluster state.